templates/tab/ts/default/src/components/sample/lib/useGraphToolkit.ts

```diff
@@ -0,0 +1,32 @@
+import { LogLevel, setLogLevel, setLogFunction } from "@microsoft/teamsfx";
+import { Providers, ProvidersChangedState, ProviderState } from "@microsoft/mgt-element";
+import { TeamsFxProvider } from "@microsoft/mgt-teamsfx-provider";
+import React from "react";
+
+export function useGraphToolkit() {
+  const [providerState, setProviderState] = React.useState<ProviderState>(ProviderState.Loading);
+      
+  React.useEffect(() => {
+    if (process.env.NODE_ENV === "development") {
+      setLogLevel(LogLevel.Verbose);
+      setLogFunction((level: LogLevel, message: string) => { console.log(message); });
+    }
+
+    Providers.globalProvider = new TeamsFxProvider({      
```

Comment:
Seems current TeamsFxProvider trigger the login pop up window automatically, this will be blocked by the browser by default, we will have error below:
![image](https://user-images.githubusercontent.com/5545529/144547971-d2b62069-0f0f-400e-861a-bc8516b06a4e.png)

It is better to have a login function, that developer can call this function when button clicked

Reply:
Isn't this the promise of Teams SSO though? I'm curious to have @waldekmastykarz thoughts on that.

Reply:
Isn't the consent separate from SSO and are we seeing here the consent fail not the SSO itself?

Reply:
The login function in teamsfx sdk is used to exchange Graph token, it will popup a login page to exchange graph access token using Teams SSO token.

From your TeamsFxProvider implementation, the login function will be called in the constructor, this will trigger the pop up window:
https://github.com/microsoftgraph/microsoft-graph-toolkit/blob/a19718867b0ff15a72b0491257720d3928bb878f/samples/teamsfx-app/tabs/src/components/TeamsFxProvider.js#L87

It is better not to call the login function in the constructor to avoid automatically popup the login page, maybe we can have a public login function in TeamsFxProvider, which can be called by developer manually in a button click callback function

Reply:
So the pattern would be to click a "Login" button when it requires consent. Would this also break in the client or this would get forced on the rich client? Could we identify when we are running in web and offer the button there, and offer the direct pop-up when running in the client?

Reply:
Yes, that's true, login button is required for consent when in browser.

I think it is developer's responsibility to determine when he want to popup login window. In our SDK, we do not need to identify whether it inside browser or inside client, SDK just need to provide a login function for developers.